HTML tidy

bastianallgeier opened the pull request:
## 👷 Status: work in progress

### Todos for this PR

- [x] @bastianallgeier Review TODO comments
  - [x] ["Escape XSS in query params"](https://github.com/getkirby/kirby/blob/feature/tidy/src/Toolkit/Dom.php#L572): Not needed
  - [x] [Attribute value escaping](https://github.com/getkirby/kirby/blob/feature/tidy/src/Toolkit/Dom.php#L587-L588): Not needed
- [x] @lukasbestle Clean up commit history
- [x] @lukasbestle Complete release notes

### Todos that can be tackled after merging

- [x] @bastianallgeier HTML sanitization in the writer field (on save)
  Must-have for 3.6.0
- [ ] @bastianallgeier Customizable sanitization/validation for any field in the blueprint (with configurable `Sane` type, `html` being the default)
  Would be great; if we do it, we should ideally do it in beta 2
- [ ] @bastianallgeier File uploads with sanitization instead of validation
  Optional, can be added later
  The idea is: In the file blueprint, there could be a new option like `sane: sanitize|validate|false`, probably with a better name. Or alternatively (or in addition) there could be a `config.php` option to define a global default for all file uploads. `validate` should always be the default if no other value is configured (`sanitize` can remove too much from files, so validation is much safer).
- [x] @lukasbestle More tests for `Sane\Html` and for `Toolkit\Dom`

## Release notes

### Breaking changes (core)

- The public static configuration attributes of the `Sane` classes have been renamed for consistency with the new `Toolkit\Dom` class
- The `Sane\Svg::$allowedNamespaces` attribute now expects namespace reference names as the array keys instead of full attribute names (so `xmlns:xlink` should become just `xlink` and `xmlns` should become an empty string)
- The `Sane\Svg::$allowedTags` attribute (before `$allowedElements`) now is an associative array with the allowed attributes for this tag as the value

### Features

- New `Toolkit\Dom` class that allows to parse, modify and sanitize DOM-based documents (XML and HTML)
- The `Sane` classes now have new `sanitize()` and `sanitizeFile()` methods that will apply the same rules as `validate()` but directly modify the input string/file so it is valid afterwards
- New `Sane\Html` class for validating and sanitizing HTML files or HTML snippets
- New `$file->sanitizeContents()` method

### Enhancements

- Added server-side sanitization in the writer field on field changes for extra security against XSS attacks
- Improved error messages for file validation/sanitization errors (`Sane` classes)
- The `Sane` classes now have proper XML namespace support so that custom namespace prefixes are properly detected

bastianallgeier committed:
New HTML tidying classes/methods

distantnative commented on src/Sane/Html.php:
```diff
@@ -0,0 +1,45 @@
+     * @param string $string
+     * @return string
+     */
+    public static function tidy(string $string): string
```
It's nit picking, but wouldn't `sanitize()` be the more fitting name next to `validate()`?

lukasbestle replied:
`Sane::sanitize()` sounds a bit duplicated though. Another alternative: `purify()`. But I'd be happy with either one of these three.

distantnative replied:
But it would be `Sane\Html::sanitize()`. Which I think reads actually quite nice. Of course there'd be `Sane\Sane::sanitize()` which is a bit quirky but so is `Sane\Sane` on its own (we have that in a few places - `Data\Data`, `Image\Image` - I think it's fine). But if you both prefer `tidy` it's fine. Was just a thought.

bastianallgeier replied:
I had sanitize first and then went for tidy, exactly for that reason. `Sane\Html::sanitize()` would be nice indeed, but we never use it that way.

lukasbestle replied:
Now that I think about it again, I do like `sanitize()`. The name is more clear as `tidy` could also be just an indentation fixer or minifier. `sanitize` has a clear connection to security.

bastianallgeier replied:
TBH I don't mind enough :) Both names work.

distantnative replied:
https://github.com/getkirby/kirby/pull/3524

lukasbestle reviewed (commented):
Unfortunately I didn't get to do a full review yet, will do so tomorrow.

First thoughts related to the general structure:
- I think the `DOM` classes should use default values in `public static` props just like the other `Sane` classes. `array_merge` has the huge disadvantage that it's impossible to pass completely different values (each one needs to be set to `false` manually). Also users cannot globally configure the allowed and disallowed elements like with the other classes.
- `denied` -> `disallowed` for consistency
- IMO the `DOM` classes should be named `Dom` for consistency with `Toolkit\Xml` and `Toolkit\Html`.
- The `tidyFile()` method should write the tidied data back to the file and not just return it. Otherwise one could use `tidy(file_get_contents())`. Not doing it in-place feels weird and unexpected to me.

Commit:
Rename to `::sanitize` (#3524)

lukasbestle committed:
Rename DOM -> Dom

lukasbestle committed:
Denied -> Disallowed

lukasbestle committed:
Sane::sanitizeFile(): Use all applying handlers

lukasbestle committed:
Sane::sanitizeFile(): Overwrite sanitized file

lukasbestle committed:
Sane: Move configuration from Dom to Html class

lukasbestle committed:
Sane: Only sanitize with one handler

https://github.com/getkirby/kirby/pull/3512#issuecomment-882246689

lukasbestle committed:
Sane: Cleanup and more unit tests

lukasbestle committed:
Overhauled Dom class

distantnative commented on src/Sane/Sane.php:
```diff
@@ -109,8 +111,23 @@ public static function sanitizeFile(string $file, $typeLazy = false): void
+                $handlerNames = array_map('get_class', $handlers);
+                throw new LogicException(
+                    'Cannot sanitize file as more than one handler applies: ' .
+                    $handlerNames
```
@lukasbestle isn't `$handlerNames` still an array here?

lukasbestle replied:
Good catch! ✅

bastianallgeier committed:
Various smaller improvements for the HTML and Dom classes

bastianallgeier committed:
Improved loop to remove nodes more reliably

lukasbestle committed:
Dom: Switch from for loops to iterator_to_array()

This gives us slightly cleaner code and also avoids regressions in the future when PHP's `DOMNodeList` implementation is changed in the future.

lukasbestle committed:
Dom: Remove duplicated logic

lukasbestle committed:
Dom: Stricter logic in isAllowedGlobalAttr()

Any invalid configuration will lead to a non-allowed attribute

lukasbestle committed:
Sane/Dom: Minor improvements

lukasbestle committed:
Sane: Consistent output from XML sanitization

lukasbestle committed:
Enhancements and tests for XML sanitization

lukasbestle committed:
`Dom::sanitize()`: New allowedAttrPrefixes option

lukasbestle committed:
Use more specific DOM classes

lukasbestle committed:
Make Dom fully namespace-aware

lukasbestle committed:
Dom: Case-insensitive matching for disallowed tags

lukasbestle committed:
Dom: Make internal helper methods public

For easier testing and for use in other classes

lukasbestle committed:
Dom: Fix logic errors

lukasbestle committed:
Cleanup

lukasbestle committed:
Dom: Always convert type argument to uppercase

lukasbestle committed:
Dom: More strict error checking

lukasbestle committed:
Sane: Adapt Svg/Svgz handlers

lukasbestle committed:
Sane: Tests for shared classes

lukasbestle committed:
Sane: Adapt Html handler

lukasbestle committed:
New Filesystem\File::sanitizeContents method

bastianallgeier committed:
Remove unnecessary escape and todo

bastianallgeier committed:
Sanitize the writer field value

Co-authored-by: Lukas Bestle <lukas@getkirby.com>

lukasbestle reviewed (approved):
I'll approve this PR for now so it can be merged if needed. I didn't finish the tests yesterday, but I don't consider them a blocker. So IMO the PR can be merged with the failing Codecov check (temporarily disable branch protection rule).
